From 1de2cef2e0aecb0c7276db06b4007c4423dd7a08 Mon Sep 17 00:00:00 2001
From: Eric Chanudet <chanudete@ainfosec.com>
Date: Mon, 9 Apr 2018 14:36:56 -0400
Subject: [PATCH] patch: Use relative path symlink to layer patches.

Uses relative path to symlink patches in tmpdir to their layer origin:
${S}/patches/a.patch -> ../../../../../../../<recipe-dir>/a.patch

This is only useful when building in a container using a file-system
binding, the absolute path in the container will likely not be valid
outside.

Signed-off-by: Eric Chanudet <chanudete@ainfosec.com>
---
 meta/lib/oe/patch.py | 9 ++++++++-
 1 file changed, 8 insertions(+), 1 deletion(-)

diff --git a/meta/lib/oe/patch.py b/meta/lib/oe/patch.py
index f1ab3dd809..44f2cacce8 100644
--- a/meta/lib/oe/patch.py
+++ b/meta/lib/oe/patch.py
@@ -610,7 +610,14 @@ class QuiltTree(PatchSet):
         if not self.initialized:
             self.InitFromDir()
         PatchSet.Import(self, patch, force)
-        oe.path.symlink(patch["file"], self._quiltpatchpath(patch["file"]), force=True)
+
+        topdir = self.d.getVar('TOPDIR')
+        dst = self._quiltpatchpath(patch["file"])
+        src = os.path.join(topdir, patch["file"])
+        rel = os.path.relpath(src, dst)
+        # Remove one '../' as relpath is called with the "file" basename at the
+        # end (see _quiltpatchpath).
+        oe.path.symlink(rel[3:], dst, force=True)
         with open(os.path.join(self.dir, "patches", "series"), "a") as f:
             f.write(os.path.basename(patch["file"]) + " -p" + patch["strippath"] + "\n")
         patch["quiltfile"] = self._quiltpatchpath(patch["file"])
-- 
2.16.1

